wasm_node_api/wasm_node_api.cc

```diff
@@ -0,0 +1,886 @@
+#include <type_traits>
+#include <functional>
+#include <uv.h>
+
+#define NODE_WANT_INTERNALS 1
+#include "js_native_api_v8.h"
+#include "wasm_node_api.h"
+#include "node_api.h"
+#include "v8-wasm.h"
+#include "env.h"
+#undef NODE_WANT_INTERNALS
+
+namespace node {
+namespace wasm_napi {
+
+namespace w = v8::wasm;
+
+using callback = w::Func::callbackType;
+
+template<typename T>
+struct is_function_pointer {
+  static const bool value =
+    std::is_pointer<T>::value ?
+      std::is_function<typename std::remove_pointer<T>::type>::value :
+      false;
+};
+
+template<typename T>
+constexpr auto is_function_pointer_v = is_function_pointer<T>::value;
+
+template<typename T>
+class Native {
```

Comment:
`Native` and `Wasm` are essentially type annotations that tell us whether a value was created in native code or in WASM code. This is important because of the different pointer sizes between systems, different memory spaces of native code vs. WASM, and other differences like function pointer representations. These types are defined such that `kind<Wasm<T>> == T` and `kind<Native<T>> == T`, allowing us to "unwrap" the types when we actually use them.

Reply:
It sounds like this should be converted to a comment in the code.